src/bigml/api/utils.clj

```diff
@@ -5,20 +5,65 @@
 (ns bigml.api.utils
   "Offers functions specific for BigML clusters.
      https://bigml.com/developers/clusters"
+  (:require [clojure.set :as set])
   (:require (bigml.api [core :as api])))
 
+(defn resource-type
+  "Returns the resource type of the specified resource.
+  The resource can be specified through:
+  - its full UUID, e.g., 'model/12234567890';
+  - a minimal map-representation, e.g., { :resource '...' };
+  - a full JSON representation."
+  [resource]
+  (let [resource (or (:resource resource) (str resource))]
+    (first (clojure.string/split resource #"/"))))
+
+(defn- get-keypath
+  "Returns the passed map m if it contains the given keypath ks,
+  otherwise nil. A keypath is a sequence of keys for recursive lookup."
+  [m ks]
+  (and (get-in m ks) m))
+
+(defn normalized-resource
+  "Makes sure the passed resource contains enough information for input
+  normalization (i.e., :input_fields and additionally :restype/field
+  when by-name is true. If it does not, then the resoure is fetched."
+  [resource by-name]
+  (let [ res-type (resource-type resource)]
+    (or (and (get-keypath resource [:input_fields])
```

Comment:
Similar comment to before, but I think `(contains? resource :input_fields)` would be more natural here.  As you said, `get-keypath` has different semantics and returns the entire object if the key is present.  But since this `and` won't ever return the first element, `contains?` is all you need.